testutil/eventually.go

```diff
@@ -0,0 +1,64 @@
+package testutil
+
+import (
+	"context"
+	"testing"
+	"time"
+)
+
+// Eventually is like require.Eventually except it takes a context.
+// If ctx times out, the test will fail.
+// ctx must have a deadline set or this will panic.
+func Eventually(ctx context.Context, t testing.TB, condition func() bool, tick time.Duration) bool {
+	t.Helper()
+
+	if _, ok := ctx.Deadline(); !ok {
+		panic("developer error: must set deadline on ctx")
+	}
+
+	ch := make(chan bool, 1)
+	ticker := time.NewTicker(tick)
+	defer ticker.Stop()
+	for tick := ticker.C; ; {
+		select {
+		case <-ctx.Done():
+			t.Errorf("Await timed out")
```

Comment:
```suggestion
			t.Errorf("Eventually timed out")
```

Reply:
I wonder if we could come up with something slightly different here? It feels like doing `require.True(t, testutil.Eventually(...))` is extra work since we're failing the test anyway, and the only acceptable return value is `true`.

We could either do `require.NoError(t, ctx.Err(), "Eventually timed out")` here and not return a boolean from `Eventually`. Or we could do `assert.NoError` and it's the callers responsibility to exit, like `if !testutil.Eventually(...) { return }`.

Also, unless we do the "no goroutine" thing I suggested below, we should also ensure that `ch` is emptied in a `t.Cleanup` (we don't have to immediately wait for it but best not to have dangling goroutines around).

Reply:
I'm hesitant to use `require` in here, as it could lead to unexpected trip-ups down the line. But returning a boolean lets callers do `require.True(testutil.Eventually(...))` which isn't much more typing than `require.Eventually`.

Reply:
Yeah that's fair, `require` definitely gives the developer less control 👍🏻.